File: tests_python/meta/test_meta_estimators.py
"""Tests for meta estimators"""
import unittest
import numpy as np
import sys
import os

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

import ingenuityml
import random

class TestMetaEstimators(unittest.TestCase):
    def setUp(self):
        rng = np.random.RandomState(42)
        self.X = rng.randn(80, 4).astype(np.float64)

        y_class = np.zeros(self.X.shape[0], dtype=np.float64)
        for i in range(self.X.shape[0]):
            cls = 0
            if self.X[i, 0] > 0:
                cls += 1
            if self.X[i, 1] > 0:
                cls += 1
            y_class[i] = cls
        self.y_class = y_class

        self.Y_multi = np.zeros((self.X.shape[0], 2), dtype=np.float64)
        self.Y_multi[:, 0] = (self.X[:, 0] > 0).astype(np.float64)
        self.Y_multi[:, 1] = (self.X[:, 1] > 0).astype(np.float64)

        self.Y_reg = np.zeros((self.X.shape[0], 2), dtype=np.float64)
        self.Y_reg[:, 0] = self.X.sum(axis=1)
        self.Y_reg[:, 1] = self.X.mean(axis=1)

    def test_one_vs_rest(self):
        model = ingenuityml.meta.OneVsRestClassifier(lambda: ingenuityml.naive_bayes.GaussianNB())
        model.fit(self.X, self.y_class)
        preds = model.predict(self.X)
        self.assertEqual(preds.shape[0], self.X.shape[0])

    def test_one_vs_one(self):
        model = ingenuityml.meta.OneVsOneClassifier(lambda: ingenuityml.naive_bayes.GaussianNB())
        model.fit(self.X, self.y_class)
        preds = model.predict(self.X)
        self.assertEqual(preds.shape[0], self.X.shape[0])

    def test_output_code(self):
        model = ingenuityml.meta.OutputCodeClassifier(lambda: ingenuityml.naive_bayes.GaussianNB(), code_size=5, random_state=42)
        model.fit(self.X, self.y_class)
        preds = model.predict(self.X)
        self.assertEqual(preds.shape[0], self.X.shape[0])

    def test_multioutput_classifier(self):
        model = ingenuityml.meta.MultiOutputClassifier(lambda: ingenuityml.naive_bayes.GaussianNB())
        model.fit(self.X, self.Y_multi)
        preds = model.predict(self.X)
        self.assertEqual(preds.shape, (self.X.shape[0], 2))

    def test_classifier_chain(self):
        model = ingenuityml.meta.ClassifierChain(lambda: ingenuityml.naive_bayes.GaussianNB())
        model.fit(self.X, self.Y_multi)
        preds = model.predict(self.X)
        self.assertEqual(preds.shape, (self.X.shape[0], 2))

    def test_multioutput_regressor(self):
        model = ingenuityml.meta.MultiOutputRegressor(lambda: ingenuityml.linear_model.LinearRegression())
        model.fit(self.X, self.Y_reg)
        preds = model.predict(self.X)
        self.assertEqual(preds.shape, (self.X.shape[0], 2))

    def test_regressor_chain(self):
        model = ingenuityml.meta.RegressorChain(lambda: ingenuityml.linear_model.LinearRegression())
        model.fit(self.X, self.Y_reg)
        preds = model.predict(self.X)
        self.assertEqual(preds.shape, (self.X.shape[0], 2))

if __name__ == '__main__':
    loader = unittest.TestLoader()
    suite = loader.loadTestsFromModule(sys.modules[__name__])

    test_methods = [test for test in suite]
    random.seed(42)
    random.shuffle(test_methods)

    shuffled_suite = unittest.TestSuite(test_methods)
    runner = unittest.TextTestRunner(verbosity=2)
    runner.run(shuffled_suite)

File: tests/meta/test_meta_estimators.cpp
#include <gtest/gtest.h>
#include "ingenuityml/meta_estimators.hpp"
#include "ingenuityml/naive_bayes.hpp"
#include "ingenuityml/linear_model.hpp"
#include <Eigen/Dense>

namespace ingenuityml {
namespace test {

class MetaEstimatorsTest : public ::testing::Test {
protected:
    void SetUp() override {
        n_samples = 80;
        n_features = 4;
        X = MatrixXd::Random(n_samples, n_features);

        y_class = VectorXd::Zero(n_samples);
        for (int i = 0; i < n_samples; ++i) {
            int cls = 0;
            if (X(i, 0) > 0) cls += 1;
            if (X(i, 1) > 0) cls += 1;
            y_class(i) = cls;
        }

        Y_multi = MatrixXd::Zero(n_samples, 2);
        for (int i = 0; i < n_samples; ++i) {
            Y_multi(i, 0) = X(i, 0) > 0 ? 1.0 : 0.0;
            Y_multi(i, 1) = X(i, 1) > 0 ? 1.0 : 0.0;
        }

        Y_reg = MatrixXd::Zero(n_samples, 2);
        for (int i = 0; i < n_samples; ++i) {
            Y_reg(i, 0) = X.row(i).sum();
            Y_reg(i, 1) = X.row(i).mean();
        }
    }

    int n_samples;
    int n_features;
    MatrixXd X;
    VectorXd y_class;
    MatrixXd Y_multi;
    MatrixXd Y_reg;
};

TEST_F(MetaEstimatorsTest, OneVsRestClassifierFitPredict) {
    auto factory = []() { return std::make_shared<naive_bayes::GaussianNB>(); };
    meta::OneVsRestClassifier clf(factory);
    clf.fit(X, y_class);
    VectorXi preds = clf.predict_classes(X);
    EXPECT_EQ(preds.size(), n_samples);
}

TEST_F(MetaEstimatorsTest, OneVsOneClassifierFitPredict) {
    auto factory = []() { return std::make_shared<naive_bayes::GaussianNB>(); };
    meta::OneVsOneClassifier clf(factory);
    clf.fit(X, y_class);
    VectorXi preds = clf.predict_classes(X);
    EXPECT_EQ(preds.size(), n_samples);
}

TEST_F(MetaEstimatorsTest, OutputCodeClassifierFitPredict) {
    auto factory = []() { return std::make_shared<naive_bayes::GaussianNB>(); };
    meta::OutputCodeClassifier clf(factory, 5, 42);
    clf.fit(X, y_class);
    VectorXi preds = clf.predict_classes(X);
    EXPECT_EQ(preds.size(), n_samples);
}

TEST_F(MetaEstimatorsTest, MultiOutputClassifierFitPredict) {
    auto factory = []() { return std::make_shared<naive_bayes::GaussianNB>(); };
    meta::MultiOutputClassifier clf(factory);
    clf.fit(X, Y_multi);
    MatrixXi preds = clf.predict(X);
    EXPECT_EQ(preds.rows(), n_samples);
    EXPECT_EQ(preds.cols(), 2);
}

TEST_F(MetaEstimatorsTest, ClassifierChainFitPredict) {
    auto factory = []() { return std::make_shared<naive_bayes::GaussianNB>(); };
    meta::ClassifierChain clf(factory);
    clf.fit(X, Y_multi);
    MatrixXi preds = clf.predict(X);
    EXPECT_EQ(preds.rows(), n_samples);
    EXPECT_EQ(preds.cols(), 2);
}

TEST_F(MetaEstimatorsTest, MultiOutputRegressorFitPredict) {
    auto factory = []() { return std::make_shared<linear_model::LinearRegression>(); };
    meta::MultiOutputRegressor reg(factory);
    reg.fit(X, Y_reg);
    MatrixXd preds = reg.predict(X);
    EXPECT_EQ(preds.rows(), n_samples);
    EXPECT_EQ(preds.cols(), 2);
}

TEST_F(MetaEstimatorsTest, RegressorChainFitPredict) {
    auto factory = []() { return std::make_shared<linear_model::LinearRegression>(); };
    meta::RegressorChain reg(factory);
    reg.fit(X, Y_reg);
    MatrixXd preds = reg.predict(X);
    EXPECT_EQ(preds.rows(), n_samples);
    EXPECT_EQ(preds.cols(), 2);
}

} // namespace test
} // namespace ingenuityml

int main(int argc, char **argv) {
    ::testing::FLAGS_gtest_shuffle = true;
    ::testing::FLAGS_gtest_random_seed = 42;

    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
